Automattic/newspack-ios
# Issues/91 syncing part 1

Refs #91 

This PR is part one of five that will introduce comprehensive syncing support to the app. 

Changes in this PR:
- Support creating a remote draft for a story folder if it needs one. 
- Support for updating a story folder's properties (postID and name) based on the remote draft.
- Support for remove story folders that have been published, marked as spam, or trashed. 
- Introduces a SyncCoordinator. This class is responsible for coordinating batch sync operations in a way that does not conflict with the Reconciler. 
- Adds a bunch of unit tests.

This PR does not: 
- Fully implement sync. The changes are already very large so more work will come in the following PRs. For now assume this is the baseline functionality, but StoryFolder syncing is expected to be incomplete.
- Do any uploading of assets.

Notes: 
- There are fewer real changes in FolderStore than shown by the diff. I moved methods around for better grouping but that code was largely untouched otherwise.   The interesting bits will be in sections for "Syncing related", "Folder Store Retrieval", and the methods for assigning a post ID, and updating sync date.
- I've opted to abandon the flex pattern for the SyncController.  Attempting to leverage it just added additional complexity with meager gain. Instead I'm relying on callback blocks and dispatch queues to ensure work is done when it needs to get done and to report back completion.  I'm not entirely satisfied with the pattern but it's at least a measure better from where Flex was going.  Open to other ideas / suggestions!

### To Test: 
Build with Xcode 11.7 (Xcode 12 support coming in the future.)
Run unit tests.  Ensure they all pass. 

From a clean install 
- Log in.
- Create a new story folder. 
- Go to the photos app and share a photo to your newly created folder.
- Return to the app and observe the console output.  Confirm that you see the sync coordinator started and stopped, and that it ran _after_ the reconciler. 
- Go to your site's and confirm you have a new draft post with the name you gave the story folder. 

Apologies in advance for size of this one!  Could I still trouble you for a peek when you have time @jleandroperez?
